---
title: "02_cor_viz"
author: "Puvvula"
date: "2024-06-25"
output: pdf_document
---

#formatting
```{r}
cor_viz<- readRDS("~/Documents/methyl_concor/result/methyl_conc_results.rds") |>
  filter(!is.na(p_value) & p_value != "",  !is.na(cor) & cor != "") |>
  mutate(
    p_value = as.numeric(p_value),
    p_value = ifelse(p_value == 0, 1.32e-14, p_value),
    ucsc_ref_gene_name = sapply(strsplit(gsub("\\s+", "", ucsc_ref_gene_name), ";"), function(x) paste(unique(x), collapse = "; ")),
    ucsc_ref_gene_group = sapply(strsplit(gsub("\\s+", "", ucsc_ref_gene_group), ";"), function(x) paste(unique(x), collapse = "; "))
  ) |>
  group_by(concord) |>
  mutate(q_value = p.adjust(as.numeric(p_value), method = "fdr")) |>
  ungroup() |>
  select(c(4,1,5,6,7,8,9,2,3,10))


# Ensure the 'chr' variable is in the correct order
chr_levels <- paste0("chr", 1:22) # Add more levels if needed, e.g., for chrX, chrY

#Unique gene group values
#unique(unlist(strsplit(gsub("[0-9]", "", cor_viz$ucsc_ref_gene_group), "; ")))
```

#density plot of p and q values - for check
```{r}
x<-cor_viz |>
  select(c(1,9,10)) |>
  pivot_longer(cols = c(p_value, q_value), names_to = "variable", values_to = "value")

ggplot(x, aes(x = value, color = variable)) +
  geom_density(alpha = 0.5) +
  labs(title = " ",
       x = "Value",
       y = "Density",
       color = "Variable") +
  theme_minimal()+
  theme(
    legend.position = "bottom",
    legend.box = "horizontal",
    legend.key = element_rect(fill = "white", color = "black"),
    legend.background = element_blank()
  )+
  facet_wrap(~ concord, scales = "free")
```


#export full results for - Appendix
```{r}
write.table(cor_viz, "~/Documents/methyl_concor/result/methylation_suppl_with_q.txt", 
            sep = "\t", row.names = FALSE, quote = FALSE)
```

#visualizing result - Figure 1 
```{r}
summary_data <- cor_viz |>
  mutate(cor_group = case_when(
           cor <= -0.75 ~ "<= -0.75",
           cor > -0.75 & cor <= -0.5 ~ "> -0.75 & <= -0.5",
           cor > -0.5 & cor <= -0.25 ~ "> -0.5 & <= -0.25",
           cor > -0.25 & cor < 0 ~ "> -0.25 & < 0",
           cor == 0 ~ "0",
           cor > 0 & cor <= 0.25 ~ "> 0 & <= 0.25",
           cor > 0.25 & cor <= 0.5 ~ "> 0.25 & <= 0.5",
           cor > 0.5 & cor <= 0.75 ~ "> 0.5 & <= 0.75",
           cor > 0.75 ~ "> 0.75"
         ),
         concord = factor(
    case_when(
      concord == "cbmc_fp" ~ "CBMC-FP",
      concord == "cbmc_mp" ~ "CBMC-MP",
      concord == "fp_mp" ~ "FP-MP",
      TRUE ~ as.character(concord)  # Keep other levels unchanged
    ),
    levels = c("CBMC-FP", "CBMC-MP", "FP-MP")  # Define the levels in the desired order
  )) |>
  group_by(concord, cor_group, ucsc_ref_gene_group, chr) |>
  summarise(count = sum(q_value < 0.05)) |>
  mutate(ucsc_ref_gene_group = sapply(strsplit(as.character(ucsc_ref_gene_group), "; "), function(parts) {
    # Check if any part contains "TSS"
    if (any(grepl("TSS", parts))) {
      "TSS"
    } else if (length(parts) > 1) {
      # If more than one part and none contains "TSS", label as "cross listed"
      "Cross listed"
    } else {
      # Otherwise, return the original value
      parts[1]
    }
  }),
  ) |>
  filter(
    cor_group == "> 0.75" |
    cor_group == "> 0.5 & <= 0.75" |
    cor_group == "> -0.75 & <= -0.5" 
  )

ggplot(summary_data, aes(x = cor_group, fill = ucsc_ref_gene_group, y = count)) +
  geom_bar(stat = "identity") +
  scale_colour_viridis_d(option = "B")+
  labs(title = " ",
       x = "Spearman correlation coefficient",
       y = "Count of CpGs (q-value<0.05)") +
  theme_minimal()+
  facet_grid(concord ~ factor(chr, levels = chr_levels), scales = "free_y", switch = "both")+
  #facet_wrap(~ concord, scales = "free", ncol=1)+
  labs(fill = "UCSC reference gene group") +
  guides(fill = guide_legend(nrow = 1, byrow = TRUE))+
  theme(legend.position = "bottom",
        axis.text.x = element_text(angle = 90, hjust = 1, vjust = 0.5),
        strip.placement = "outside",
        axis.title = element_text(color = "black", size = 11),
        axis.text = element_text(color = "black", size = 11),
        strip.text = element_text(color = "black", size = 11),
        panel.grid.minor.x = element_blank(),
        panel.grid.major.x = element_blank())

ggsave("~/Documents/methyl_concor/result/correlation_summary.tiff",
       width = 12.5,height = 6.5,
       dpi=300,
       bg="white"
       )
```

#function for cross-listing issue
```{r}
extract_and_sum_after_colon <- function(text) {
  # Extract the part within parentheses
  within_parens <- str_extract(text, "\\((.*?)\\)")
  
  if (is.na(within_parens)) {
    return(NA_real_)
  }
  
  # Extract numeric values that follow ": "
  numbers <- str_extract_all(within_parens, "(?<=:\\s)\\d+")
  
  # Convert to numeric and sum
  numbers <- as.numeric(unlist(numbers))
  
  # Return the sum of numeric values
  return(sum(numbers, na.rm = TRUE))
}
```

#updated table for manuscript - Supplement table 1
```{r}
summary <- cor_viz |>
  #filter(p_value < 0.05) |>
  filter(q_value < 0.05 
         & cor > 0.85
         )|>
  mutate(p_value_group = ifelse(p_value < 0.05 , "< 0.05", "> 0.05"),
         q_value_group = ifelse(q_value < 0.05 , "< 0.05", "> 0.05"),
         cor_group = case_when(
           cor <= -0.75 ~ "<= -0.75",
           cor > -0.75 & cor <= -0.5 ~ "> -0.75 & <= -0.5",
           cor > -0.5 & cor <= -0.25 ~ "> -0.5 & <= -0.25",
           cor > -0.25 & cor < 0 ~ "> -0.25 & < 0",
           cor == 0 ~ "0",
           cor > 0 & cor <= 0.25 ~ "> 0 & <= 0.25",
           cor > 0.25 & cor <= 0.5 ~ "> 0.25 & <= 0.5",
           cor > 0.5 & cor <= 0.75 ~ "> 0.5 & <= 0.75",
           cor > 0.75 ~ "> 0.75"
         )) |>
  group_by(concord, q_value_group, cor_group, chr) |>
  summarise(
    count_cpg_sites = n(),
    count_genes = n_distinct(ucsc_ref_gene_name[ucsc_ref_gene_name != ""]),
    gene_names = paste(unique(ucsc_ref_gene_name[ucsc_ref_gene_name != ""]), collapse = ", "),
    cross_listed_genes = sum(sapply(str_split(ucsc_ref_gene_name, ";"), function(x) length(unique(x)) > 1 && length(unique(x)) == length(x))),
    chr_cross_listed = paste0(
      unique(chr) %>% factor(levels = chr_levels) %>% sort() %>% as.character(),
      " (",
      sapply(unique(chr) %>% factor(levels = chr_levels) %>% sort() %>% as.character(), function(x) paste(
        c(
          if (sum(chr == x & str_detect(ucsc_ref_gene_group, "TSS")) > 0) paste0("TSS: ", sum(chr == x & str_detect(ucsc_ref_gene_group, "TSS"))),
          if (sum(chr == x & str_detect(ucsc_ref_gene_group, "Body")) > 0) paste0("Body: ", sum(chr == x & str_detect(ucsc_ref_gene_group, "Body"))),
          if (sum(chr == x & str_detect(ucsc_ref_gene_group, "3'UTR")) > 0) paste0("3UTR: ", sum(chr == x & str_detect(ucsc_ref_gene_group, "3'UTR"))),
          if (sum(chr == x & str_detect(ucsc_ref_gene_group, "5'UTR")) > 0) paste0("5UTR: ", sum(chr == x & str_detect(ucsc_ref_gene_group, "5'UTR"))),
          if (sum(chr == x & str_detect(ucsc_ref_gene_group, "Exon")) > 0) paste0("Exon: ", sum(chr == x & str_detect(ucsc_ref_gene_group, "Exon"))),
          if (sum(chr == x & ucsc_ref_gene_group == "") > 0) paste0("NA: ", sum(chr == x & ucsc_ref_gene_group == ""))
        ),
        collapse = ", "
      )),
      ")"
    ) |>
      paste(collapse = "; "),
    cpg_cross_listed_regions = sum(str_detect(ucsc_ref_gene_group, "TSS") + str_detect(ucsc_ref_gene_group, "Body") + str_detect(ucsc_ref_gene_group, "3'UTR") + str_detect(ucsc_ref_gene_group, "5'UTR") > 1),
    numeric_sum = sapply(chr_cross_listed, extract_and_sum_after_colon)
  )

write_csv(summary,
          file="~/Documents/methyl_concor/result/concordance_summary_with_q.csv")
```

#for manuscript text
```{r}
summary_txt <- cor_viz |>
  #filter(p_value < 0.05) |>
  filter(q_value < 0.05)|>
  mutate(p_value_group = ifelse(p_value < 0.05 , "< 0.05", "> 0.05"),
         q_value_group = ifelse(q_value < 0.05 , "< 0.05", "> 0.05"),
         cor_group = case_when(
           cor <= -0.75 ~ "<= -0.75",
           cor > -0.75 & cor <= -0.5 ~ "> -0.75 & <= -0.5",
           cor > -0.5 & cor <= -0.25 ~ "> -0.5 & <= -0.25",
           cor > -0.25 & cor < 0 ~ "> -0.25 & < 0",
           cor == 0 ~ "0",
           cor > 0 & cor <= 0.25 ~ "> 0 & <= 0.25",
           cor > 0.25 & cor <= 0.5 ~ "> 0.25 & <= 0.5",
           cor > 0.5 & cor <= 0.75 ~ "> 0.5 & <= 0.75",
           cor > 0.75 ~ "> 0.75"
         )) |>
  group_by(concord, q_value_group, cor_group, chr) |>
  summarise(
    count_cpg_sites = n(),
    count_genes = n_distinct(ucsc_ref_gene_name[ucsc_ref_gene_name != ""]),
    cross_listed_genes = sum(sapply(str_split(ucsc_ref_gene_name, ";"), function(x) length(unique(x)) > 1 && length(unique(x)) == length(x))),
    chr_cross_listed = paste0(
      unique(chr) %>% factor(levels = chr_levels) %>% sort() %>% as.character(),
      " (",
      sapply(unique(chr) %>% factor(levels = chr_levels) %>% sort() %>% as.character(), function(x) paste(
        c(
          if (sum(chr == x & str_detect(ucsc_ref_gene_group, "TSS")) > 0) paste0("TSS: ", sum(chr == x & str_detect(ucsc_ref_gene_group, "TSS"))),
          if (sum(chr == x & str_detect(ucsc_ref_gene_group, "Body")) > 0) paste0("Body: ", sum(chr == x & str_detect(ucsc_ref_gene_group, "Body"))),
          if (sum(chr == x & str_detect(ucsc_ref_gene_group, "3'UTR")) > 0) paste0("3UTR: ", sum(chr == x & str_detect(ucsc_ref_gene_group, "3'UTR"))),
          if (sum(chr == x & str_detect(ucsc_ref_gene_group, "5'UTR")) > 0) paste0("5UTR: ", sum(chr == x & str_detect(ucsc_ref_gene_group, "5'UTR"))),
          if (sum(chr == x & str_detect(ucsc_ref_gene_group, "Exon")) > 0) paste0("Exon: ", sum(chr == x & str_detect(ucsc_ref_gene_group, "Exon"))),
          if (sum(chr == x & ucsc_ref_gene_group == "") > 0) paste0("NA: ", sum(chr == x & ucsc_ref_gene_group == ""))
        ),
        collapse = ", "
      )),
      ")"
    ) |>
      paste(collapse = "; "),
    cpg_cross_listed_regions = sum(str_detect(ucsc_ref_gene_group, "TSS") + str_detect(ucsc_ref_gene_group, "Body") + str_detect(ucsc_ref_gene_group, "3'UTR") + str_detect(ucsc_ref_gene_group, "5'UTR") > 1),
    numeric_sum = sapply(chr_cross_listed, extract_and_sum_after_colon)
  )
```

	

#manual check
```{r}
subset<- cor_viz |>
  filter(
    concord == "cbmc_fp",
    chr == "chr3",
    q_value < 0.05,
    #cor >= -0.5
    cor > -0.5 & cor <= -0.25 
    #cor >= 0.6 & cor < 0.8
  )

#for manuscript text
result <- subset %>%
  separate_rows(ucsc_ref_gene_name, sep = ";") %>%
  mutate(ucsc_ref_gene_name = str_trim(ucsc_ref_gene_name)) %>%
  filter(ucsc_ref_gene_name != "") %>%
  count(ucsc_ref_gene_name) %>%
  summarise(
    unique_values = paste(ucsc_ref_gene_name, collapse = ", "),
    total_count = n()
  )

x<- summary |>
  filter(
    concord == "fp_mp",
    cor_group == "> 0.8"
  )
```


#summary for manuscript -Supplement table 2
```{r}
cpg_sites_all_levels <- cor_viz_df %>%
  group_by(ucsc_ref_gene_name, chr, cpg_site) %>%
  summarise(levels_present = n_distinct(concord), .groups = 'drop') %>%
  filter(levels_present == 3) %>%
  group_by(ucsc_ref_gene_name, chr) %>%
  summarise(cpg_sites = paste(unique(cpg_site), collapse = ", "))
```

#manuscript last paragraph
```{r}
cor_viz_txt<- cor_viz  |>
  filter(q_value < 0.05) |>
  select(c(1,3,8)) |>
  pivot_wider(
    names_from = concord,
    values_from = cor 
  )

x<- cor_viz |>
  filter(if_all(c(3,4), ~ . >= 0.85))

#print unique gene names with high correlation
print(paste(unique(na.omit(x$ucsc_ref_gene_name[x$ucsc_ref_gene_name != ""])), collapse = ", "))
```



#ternary plot for publication
```{r}
plot_df<- cor_viz |>
  select(c(1,2,4,5,8)) |>
  pivot_wider(names_from = concord, values_from = cor) |>
  rename(`CBMC-FP` = cbmc_fp,
         `CBMC-MP` = cbmc_mp,
         `FP-MP` = fp_mp)


ggtern(data= plot_df,
       aes(x=`CBMC-FP`, y=`CBMC-MP`, z=`FP-MP`))+
  geom_point(size=0.001, alpha = 1/15)+
  stat_density_tern(geom = "polygon", 
                    color = "white",
                    alpha = 0.001) +
  theme_rgbw()+
  labs(x = "CBMC-FP", y = "CBMC-MP", z = "FP-MP")

ggsave("~/Documents/methyl_concor/result/concor_plt.tiff",
       width = 7.5,height = 7.5,
       dpi=300
       )
```


#additional figures from original data
```{r}
load(paste0(input, "/methylation.Rds"))

#12 CpGs with high correlation
cpg_sites <- c("cg15071166", "cg00864916", "cg08923669", "cg04842962", 
               "cg26022684", "cg26654770", "cg02622647", "cg10070864", 
               "cg13473803", "cg15969227", "cg22337407", "cg25602718")

filtered_methylation <- all_methylation |>
  filter(cpg_site %in% cpg_sites) |>
  select(-c(4)) |>
  mutate(gene_name = case_when(
    cpg_site == "cg15071166" ~ "CAMMK1",
    cpg_site == "cg00864916" ~ "KIAA1614",
    cpg_site == "cg08923669" ~ "MRPL28",
    cpg_site == "cg04842962" ~ "MRPS18A",
    cpg_site == "cg26022684" ~ "NBPF1",
    cpg_site == "cg26654770" ~ "NINJ2",
    cpg_site %in% c("cg02622647", "cg10070864", "cg13473803", 
                    "cg15969227", "cg22337407", "cg25602718") ~ "Not Annotated",
    TRUE ~ NA_character_  # Keep other rows as NA if no match
  ),
  smpl = str_to_upper(smpl),
  methylation_intensity = scales::rescale(methylation_intensity, to = c(0, 1))
  )

#heatmap
ggplot(filtered_methylation, aes(x = smpl, y = participant_id, fill = methylation_intensity)) +
  geom_tile(width = 1, height = 1) +
  scale_fill_viridis_c(option = "D", limits = c(0, 1),
                       labels = scales::label_number(accuracy = 0.1)) +  
  labs(x = "CpG Site", y = "Study participants", fill = "Methylation Î²") +
  theme_void() +
  theme(axis.text.x = element_text(angle = 90, hjust = 1))+
  facet_grid(.~factor(cpg_site, levels = cpg_sites)+gene_name, 
             switch = "x")+
  theme(
    axis.text.y = element_blank(),  
    axis.title.y = element_blank(), 
    legend.position = "bottom",      
    legend.box = "horizontal",
    strip.text.x = element_text(size = 11, color = "black", angle = 0),
    strip.placement = "outside",
    legend.title = element_text(margin = margin(b = 12))
  )

ggsave("~/Documents/methyl_concor/result/methylation_beta_cpg_subset.tiff",
       width = 14,height = 6,
       dpi=300,
       bg="white"
       )
```




